### DISCUSSION TITLE
Java formatting options provided as initializationOptions.settings.java.format.settings.url are ignored

### DISCUSSION THREAD

**Discussion by HaraldKi** [joaotavora/eglot#1377] [Q&A] [answered]
I wrote already a bug report to eclipse.jdt.ls, mostly to start documenting this somewhat: https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/3108 , but I get the impression that I misunderstand a concept of configuration of eglot or it may actually be a problem with eglot.

My goal: tell the language server how to format Java.

Somewhat inspired from https://github.com/joaotavora/eglot/discussions/888 I do this:

```lisp
(with-eval-after-load 'eglot
  (add-to-list
   'eglot-server-programs
   `(java-mode
     "jdtls"
     :initializationOptions
     (:settings (:java
                 (:format
                  (:enabled t :profile: "user-variant" :settings (:url "/home/user/.emacs.d/eclipse-formatter.xml"))
                  )
                 ))
     )
   )
  )
```
- In the *EGLOT...* buffer I see that these settings are sent to the server
- By adding debug logs to the server, I see that it actually reads and understands the formatter options, all looks dandy. But then eglot does this:
```
[jsonrpc] e[16:25:17.564] --> textDocument/didOpen {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument ... [clipped]
[jsonrpc] e[16:25:17.626] --> workspace/didChangeConfiguration {"jsonrpc":"2.0","method":"workspace/didChangeConfiguration","params":{"settings":{}}}
```
The `didOpen` line is only for context. There is nothing in between. The culprit are the empty `"settings"` which kills all the nice setup that was sent at initialization time. Basically eclipse.jdt.ls reacts to the empty settings with setting everything back to default.

By setting `eglot-connect-hook` to `nil`, because this seems to be the initiator of the `didChangeConfiguration`, I got what I want, formatting according to the rules provided during initialization.

Question: Am I missing some additional configuration to tell eglot to send the whole `java-mode` configuration from the initialization phase again. Or is this a misunderstanding between eglot and the server, where the former assumes `didChangeConfiguration` sends only changes whereas the server assumes it gets the full set of configurations?

Reading https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeConfiguration it says:
```javascript
interface DidChangeConfigurationParams{
	/**
	 * The actual changed settings
	 */
	settings: LSPAny;
}
```
which could be interpreted as saying: if we sent empty, nothing was changed. Yet the ecplise.jdt.ls takes this as a request to "clear" all settings (back to default).
I'll add these findings to the server issue too, but if someone feels strongly about which interpretation is correct and why I would be grateful for a statement.

**Comment by HaraldKi** [comment]
> jdt.ls deviates from the spec here.
https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/3108#issuecomment-2018589030

So I have to live with
```lisp
(setq eglot-connect-hook nil)
```
which works at least at startup. After configuration changes I need to restart the server then. I hope this is workable for a while.

I'll close this sometime soon or anyone with the keys to the empire close it if there is nothing more to say.

**Reply by manuel-uberti** [reply, marked as answer]
Are you doing that `setq` globally? I am asking because it will set `eglot-connect-hook` for every server if you do it that way. 

If that is a problem for you, I suggest something along the following lines:
```emacs-lisp
(defun my-do-not-signal-didChangeConfiguration ()
  (remove-hook 'eglot-connect-hook #'eglot-signal-didChangeConfiguration t))

(add-hook 'java-mode-hook #'my-do-not-signal-didChangeConfiguration)
```

**Reply by HaraldKi** [reply]
Shortly I want to connect to the typescript compiler too and I was already wondering how to avoid the global setting then. So great hint, as it saves me some digging.

**Comment by cayhorstmann** [comment]
You can also consider using a global workspace configuration. Then do not remove the eglot-signal-didChangeConfiguration hook! See https://horstmann.com/unblog/2025-09-20/
